src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -652,15 +660,130 @@ int PeerReplayer::remote_mkdir(const std::string &epath, const struct ceph_statx
   return 0;
 }
 
+PeerReplayer::FileTransferHandlerThreadPool::FileTransferHandlerThreadPool(
+    int _num_threads, int _max_concurrent_large_file,
+    uint64_t _file_size_threshold)
+    : num_threads(_num_threads), max_concurrent_large_file(std::min(
+                                     _num_threads, _max_concurrent_large_file)),
+      file_size_threshold(_file_size_threshold), workers(_num_threads),
+      stop_flag(true) {
+  max_token_sum = file_size_threshold * num_threads * max_concurrent_large_file;
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::activate() {
+  ceph_assert(stop_flag.load());
+  stop_flag = false;
+  for (auto &worker : workers) {
+    worker = std::thread(&FileTransferHandlerThreadPool::run, this);
+  }
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::deactivate() {
+  ceph_assert(!stop_flag.load());
+  stop_flag = true;
+  pick_task.notify_all();
+  give_task.notify_all();
+  exec_task.notify_all();
+  for (auto &worker : workers) {
+    if (worker.joinable()) {
+      worker.join();
+    }
+  }
+  drain_queue();
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::drain_queue() {
+  std::scoped_lock lock(pmtx);
+  while (!task_queue.empty()) {
+    auto& [task, token] = task_queue.front();
+    task(true);
+    current_token_sum -= token;
+    task_queue.pop();
+  }
+}
+
+// K * L + (S - L) * min(file_size, K)
+uint64_t
+PeerReplayer::FileTransferHandlerThreadPool::get_token(uint64_t file_size) {
+  return file_size_threshold * max_concurrent_large_file +
```

Comment:
```
S = limit for small transfers
L = limit for large transfers
K = file size cap (1 object)
M = max token capacity of the queue
E = tokens for file existence
P = tokens per full object

Accounted tokens for a file: E + P * min(file_size, K)/K

Case 1: only small files

S * E = M

Case 2: only large files

L * (E + P) = M

Arbitrarily set M (e.g., to 1 million)

Solve for E:  E = M / S
Solve for P:  P = M / L - E

Set token capacity of the queue to M = K * S * L
Accounted tokens for a file: K * L + (S - L) * min(file_size, K)
```
Here is the math. This math is done by @patrakov and seems convincing to me.

Reply:
Sorry, I forgot to remove one confusing remark from the above: `(e.g, to 1 million)`. Correct version: `set M to K * S * L to avoid non-integer math`. You can check the math as follows: if cephfs-mirror is asked to transfer very small (e.g, zero-byte) files, the accounted tokens would be `K * L` and so `S` files will fit in the queue. Analogously, when the file size is K, then the expression for the accounted tokens per file simplifies to `K * S` and therefore `L` files fit in the queue. The rest is just some smooth behavior in between.

What the text above misses is the justification for the whole business with different limits for large and small files. So, here it is.

The implicit assumption here is that the destination cluster is not the bottleneck. This is indeed the case even if the servers in the destination cluster have much worse specs than the source one, as the user is effectively required to treat the destination cluster as read-only; thus, there will be no background load, which I found to be a limiting factor on the source cluster.

So, we have to consider two bottlenecks: the IOPS limit for a single-threaded workload due to HDDs in the source cluster, and network bandwidth. With a single thread, different bottlenecks are hit and thus different customer complaints are provoked:

* Small files: "your old cephfs-mirror can only transfer 100 files per second, that's way too slow, please parallelize way way more" - but with too much parallelism, the HDDs in the source cluster would become overloaded. So, the small-files concurrency limit prevents the HDD overload in the source cluster.
* Large files: "your old cephfs-mirror is almost good, but please parallelize it by a factor of 2 so that it can better fill the network" - if we parallelize too much, we'll overload the network and provoke complaints about write latency on the destination cluster.

In the intended cluster for deployment, I will probably set the small-files concurrency to 64 or 128 and large-file concurrency to 4.

Reply:
> This means one will get just 2 files transferred in parallel. While I'd expect we;d like to have 3 transfers...

The limit is indeed reached for zero-sized files only, and that's intentional. Maybe we should rename the setting.

I would be open to adding some floor to the file size, but the problem is, there is no good value for such a floor, and I don't want to breed tunable parameters. So I would prefer documenting that the limit applies to zero-sized files, and that even one byte would subtract 1 from the concurrency.

Reply:
OK, this is the minimal fix of the math to address the "one-byte files only reach the `S - 1` concurrency level" objection. The idea is to aim for `S + 1` or `L + 1` concurrency but stop one token short.

```
Set token capacity of the queue to M = K * (S + 1) * (L + 1) - 1
Accounted tokens for a file: K * (L + 1) + (S - L) * min(file_size, K)
```

Reply:
Isn't it something like that we are considering 1 byte as the minimal byte size for a file. So we want to run S, 1 byte file parallel and L, K byte file in parallel, then shouldn't be it just simply,
```
Set token capacity of the queue to M = L * S * (K - 1)
Accounted tokens for a file: K * L - S + min(file_size, K) * (S - L)
```

Reply:
> Isn't it something like that we are considering 1 byte as the minimal byte size for a file. So we want to run S, 1 byte file parallel and L, K byte file in parallel, then shouldn't be it just simply,
> 
> ```
> Set token capacity of the queue to M = L * S * (K - 1)
> Accounted tokens for a file: K * L - S + min(file_size, K) * (S - L)
> ```

Well, my proposal is a more aggressive. Your proposal would apply the parallelism of S threads only to 1-byte files, but balk at 2-byte files, which would hit the same objection. My formula would extend the use of S threads to files as "big" as it is possible with a linear formula.

Reply:
I understood now. Idea is to set a lower bound for number of small files to be transferred such that if they are (K-1) length bytes, even then it should transfer >= S files

Reply:
Just trying to understand with your new formula does it transfer S, (k - 1) size byte file concurrently?

Reply:
I don't understand the question. Let me try to explain the formula in a different way.

The formula is:

```
Set token capacity of the queue to M = K * (S + 1) * (L + 1) - 1
Accounted tokens for a file: K * (L + 1) + (S - L) * min(file_size, K)
```

Let's suppose that somebody tries to transfer large (`K` bytes) files. Then, if an attempt is made to transfer `L + 1` files at once, this would consume `(L + 1) * (K * (L + 1) + (S - L) * K)` tokens, which simplifies down to `(L + 1) * K * (S + 1)`. But the queue can only contain K * (S + 1) * (L + 1) - 1 tokens, i.e., one token too few, so the `L + 1`st transfer does not happen, i.e., the concurrency for `K`-byte sized files is still `L`, as designed.

Now consider zero-sized files. Assume that one attempts to transfer  `S + 1` files at once. Then, the total number of tokens is `(S + 1) * K * (L + 1)`, which is, again, one token more than what can fit in the queue. So, the achievable concurrency is `S`, not `S + 1`.

And for the one-byte files, although the formula is not really designed for them, let's consider `S` transfers. Then, the total number of used tokens is `S * K * (L + 1) + S - L`, which is not that nice. Let's see how many free tokens remain in the queue:

```
   K * (S + 1) * (L + 1) - 1 - (S * K * (L + 1) + S - L)
= K * S * L + K * S + K * L + K - 1 - K * S * L - K * S - S + L
= K * L + K + L - 1 - S
= (K + 1) * (L + 1) - S - 2
```

This is non-negative if `S <= (K + 1) * (L + 1) - 2`. For realistic numbers, e.g., K = 4194304, S = 128, L = 4, this is of course positive, and therefore, for 1-byte files, there will be `S` concurrent transfers.

Another question is how large the "small" file can still be while still allowing `S` concurrent transfers. The amount of free tokens is:

```
   K * (S + 1) * (L + 1) - 1 - S * (K * (L + 1) + (S - L) * file_size)
= K * S * L + K * S + K * L + K - 1 - K * S * L - K * S - S * (S - L) * file_size
= K * L + K - 1 - S * (S - L) * file_size
```

So the critical file_size on the lower end is:

```
   (K * L + K - 1) / S / (S - L)
```

Which, for the example above, is 1321 bytes.

Now, consider the concurrency when transferring files of size `K - 1`. Conjecture: the concurrency for `S > L` in realistic scenarios would be `L + 1`. Indeed, the number of free tokens is then:

```
   K * (S + 1) * (L + 1) - 1 - (L + 1) * (K * (L + 1) + (S - L) * (K - 1))
= (L + 1) * (S - L) - 1
```
... which is non-negative if both `L + 1` and `S - L` are positive integers.